node/src/components/storage.rs

```diff
@@ -1021,11 +1069,110 @@ impl Storage {
             return Err(LmdbExtError::BlockHeaderNotStoredUnderItsHash {
                 queried_block_hash: *block_hash,
                 found_block_header_hash,
+                block_header: Box::new(block_header),
             });
         };
         Ok(Some(block_header))
     }
 
+    /// Retrieves a single Merklized block body in a separate transaction from storage.
+    fn get_single_v2_block_body<Tx: Transaction>(
+        &self,
+        tx: &mut Tx,
+        block_body_hash: &Digest,
+    ) -> Result<Option<BlockBody>, LmdbExtError> {
+        let [hashed_deploy_hashes, hashed_transfer_hashes, hashed_proposer]: [Digest;
+            BlockBodyHashedParts::NUMBER_OF_HASHES] =
```

Comment:
I think we'll always have to have a database like this with exactly three fields.

Here's my reasoning: eventually, we are going to update the block body.  The fast sync code is supposed to be invariant to such a change. But the fast sync code must be able to get the deploy and transfer hashes (+proof) to avoid replay attacks. So we'll need a table like this to save those.

Reply:
Hmm, why always exactly three fields? Will this work if we add more fields to the block?

I wonder whether we should be doing it differently... Have a database which would hold entries like `hash → (hash1, hash2)`. `hash1` would be the hash of a field, and `hash2` would be the Merkle proof for the rest of the block. So we'd have:

- `body_hash → (deploy_hashes_hash, hash0)`
- `hash0 → (transfer_hashes_hash, hash1)`
- `hash1 → (proposer_hash, SENTINEL)`

If we added a field, the last one would become `hash1 → (proposer_hash, hash2)` and `hash2` would point to another pair that would contain the sentinel. The field hashes would be pointers into the respective databases. This way the only thing that would change if we add fields would be the number of the databases, and storing old blocks alongside new ones would be no problem - we'd always be able to recognize what format of a block we're reading by where the sentinel appears.

What do you think?

Reply:
@fizyk20 I went ahead an implemented your proposal in https://github.com/casper-network/casper-node/pull/1563/commits/7388d0d0c843d82ca38d18d7c8a03b7860af7fd6